bothan-coingecko/src/worker/asset_worker.rs

```diff
@@ -0,0 +1,142 @@
+use std::sync::Weak;
+use std::time::Duration;
+
+use chrono::NaiveDateTime;
+use rust_decimal::prelude::FromPrimitive;
+use rust_decimal::Decimal;
+use tokio::time::{interval, timeout, Interval};
+use tracing::{info, warn};
+
+use bothan_core::store::Store;
+use bothan_core::types::AssetInfo;
+
+use crate::api::rest::CoinGeckoRestAPI;
+use crate::api::types::{Market, Order};
+use crate::worker::error::ParseError;
+use crate::worker::CoinGeckoWorker;
+
+pub(crate) fn start_asset_worker(
+    weak_worker: Weak<CoinGeckoWorker>,
+    mut update_interval: Interval,
+    page_size: usize,
+    page_query_delay: Duration,
+) {
+    tokio::spawn(async move {
+        while let Some(worker) = weak_worker.upgrade() {
+            info!("updating asset info");
+
+            let ids = worker.store.get_query_ids().await;
+
+            let result = timeout(
+                update_interval.period(),
+                update_all_asset_info(&worker.store, &worker.api, ids, page_size, page_query_delay),
+            )
+            .await;
+
+            if result.is_err() {
+                warn!("updating interval exceeded timeout")
+            }
+
+            update_interval.tick().await;
+        }
+
+        info!("worker has been dropped, stopping asset worker");
+    });
+}
+
+async fn update_all_asset_info(
+    store: &Store,
+    api: &CoinGeckoRestAPI,
+    ids: Vec<String>,
+    page_size: usize,
+    page_query_delay: Duration,
+) {
+    let mut interval = interval(page_query_delay);
+    let queue = Vec::from_iter(ids.chunks(page_size));
+    for ids in queue {
+        update_asset_info(store, api, ids, page_size).await;
+        interval.tick().await;
+    }
+}
+
+async fn update_asset_info(
+    store: &Store,
+    api: &CoinGeckoRestAPI,
+    ids: &[String],
+    page_size: usize,
+) {
+    match api
+        .get_coins_market(ids, Some(Order::IdDesc), Some(page_size), Some(1))
+        .await
+    {
+        Ok(markets) => {
+            let to_set = markets
```

Comment:
Should we sanity check that the number of markets we get is >= the id size? better to know if sth unexpected happen fast